core/ir/riscv64/opcode_api.h

```diff
@@ -0,0 +1,89 @@
+/* **********************************************************
+ * Copyright (c) 2022 Rivos, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Rivos, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL RIVOS, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef _DR_IR_OPCODES_RISCV64_H_
+#define _DR_IR_OPCODES_RISCV64_H_ 1
+
+/****************************************************************************
+ * OPCODES
+ */
+/**
+ * @file dr_ir_opcodes_arm.h
+ * @brief Instruction opcode constants for RISC-V.
+ */
+/** Opcode constants for use in the instr_t data structure. */
+enum {
+    /*   0 */ OP_INVALID,
+    /* NULL, */ /**< INVALID opcode */
+    /*   1 */ OP_UNDECODED,
+    /* NULL, */ /**< UNDECODED opcode */
+    /*   2 */ OP_CONTD,
+    /* NULL, */ /**< CONTD opcode */
+    /*   3 */ OP_LABEL,
+    /* NULL, */ /**< LABEL opcode */
+
+    /* FIXME i#3544: Put real opcodes here ans make sure ordering matches HW */
+    /* FIXME i#3544: should load/store width be separate opcode or in the instruction
```

Comment:
Trying to answer this question: does the Risc-v ISA provide separate opcodes to move different widths? If not, then we can probably just encode the width in the operand sizes.

Reply:
The load width is encoded in bits 14:12 of the instruction, the opcode is the same for all integer loads. Only floating point loads have a different opcode but they also store the width of the operand (single, double, quad) in the 14:12 bits. I'll update the comment to clarify that and add `OP_fl / OP_fs` for floating points.

Reply:
I see there are LW, LH, LB etc to load different sized values from memory. Are these just aliases for OP_l?

Reply:
Yes. `L[BHWD]` are encoded in the I-format and have the OPCODE field (bits <6:0>) set to 0000011 and width encoded in the funct3 field (bits <14:12>) as follows:
LB - 000
LH - 001
LW - 010
LD - 011

Stores and FP loads/stores use similar approach.

Reply:
Thanks.